pkg/workloads/types/reservation.go

```diff
@@ -701,31 +701,17 @@ const (
 
 //ReservationPushSignature push signature to reservation
 func ReservationPushSignature(ctx context.Context, db *mongo.Database, id schema.ID, mode SignatureMode, signature generated.SigningSignature) error {
-
+	// this function just push the signature to the reservation array
+	// there are not other checks involved here. So before calling this function
+	// we need to ensure the signature has the rights to be pushed
 	var filter ReservationFilter
 	filter = filter.WithID(id)
 	col := db.Collection(ReservationCollection)
-	// NOTE: this should be a transaction not a bulk write
-	// but i had so many issues with transaction, and i couldn't
-	// get it to work. so I used bulk write in place instead
-	// until we figure this issue out.
-	// Note, the reason we don't just use addToSet is the signature
-	// object always have the current 'time' which means it's a different
-	// value than the one in the document even if it has same user id.
-	_, err := col.BulkWrite(ctx, []mongo.WriteModel{
-		mongo.NewUpdateOneModel().SetFilter(filter).SetUpdate(
-			bson.M{
-				"$pull": bson.M{
-					string(mode): bson.M{"tid": signature.Tid},
-				},
-			}),
-		mongo.NewUpdateOneModel().SetFilter(filter).SetUpdate(
-			bson.M{
-				"$addToSet": bson.M{
-					string(mode): signature,
-				},
-			}),
-	}, options.BulkWrite().SetOrdered(true))
+	_, err := col.UpdateOne(ctx, filter, bson.M{
+		"$push": bson.M{
```

Comment:
Are u sure $push can work on a `nil` array in mongo ? Also because there is no race grantees, I will also change where we check the signature count to make sure a user is not counted twice if somehow 2 signtaures of the same user managed to end up in that array (imagine pushing the same signtuare twice but at exact same time)

Reply:
I'm still not sure where the nill array issue comes from. I'm suspecting the migration code did that.